[wip] Install Grott proxy server

andrewfg opened the pull request:
Resolves #2043

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

andrewfg commented:
@mstormi please feel free to start making comments -- even though it is still wip..

mstormi commented:
Thanks! Will review as soon as I find time. 

See action tab/builds for an image or trigger build there but that's overkill, I'm if the BATS test is.

DCO no worries 

docs yeah dunno myself maybe add a single sentence in docs/openhabian.md in the section with nginx, FIND etc. 

window size it's one of the number arguments at the end of the whiptail cmd eventually both rtfm I'm on the phone and don't recall atm.

andrewfg commented:
> add a single sentence in docs/openhabian.md in the section with nginx, FIND etc.

Done.

> one of the number arguments at the end of the whiptail cmd eventually both rtfm 

Done.

mstormi commented on functions/menu.bash:
```diff
@@ -110,7 +110,7 @@ show_main_menu() {
 
   elif [[ "$choice" == "20"* ]]; then
-    choice2=$(whiptail --title "openHABian Configuration Tool — $(get_git_revision)" --menu "Optional Components" 24 118 16 --cancel-button Back --ok-button Execute \
+    choice2=$(whiptail --title "openHABian Configuration Tool — $(get_git_revision)" --menu "Optional Components" 33 118 25 --cancel-button Back --ok-button Execute \
```
don't enlarge the window beyond 25 but use scrolling. the other argument IIRC is how many lines are shown *when scrolling*

andrewfg replied:
> don't enlarge the window beyond 25 but use scrolling

The first argument is the total heigh of the dialog, and the third is the height of the menu within it.

Currently the dialog looks as follows which shows options 21 thru 2C .. so I worry that a simple user may not see the scrollbar and may therefore not realise that 2D, 2E, and now 2F are also available. So my PR increases the whole dialog height by 9 lines and the scrolling part by 9 lines so that all options will be visible without any scrolling. But this is ultimately a style issue, so if you want to keep it as it is, that is Ok with me. (But maybe add a text "Optional Components (scroll down for more)")??

<img width="1209" height="556" alt="image" src="https://github.com/user-attachments/assets/5b655de0-8d44-476e-acc4-6dadb56be899" />

andrewfg replied:
I don't think you need to change anything other than what I did already. Here is how it looks..

<img width="959" height="550" alt="image" src="https://github.com/user-attachments/assets/9454cce3-ddbb-4281-8378-405e2fefa408" />

andrewfg replied:
I reverted the dialog box size to prior. But added the text "scroll to see more"..

mstormi commented on functions/grott-install.bash:
```diff
@@ -0,0 +1,152 @@
+    fi
+
+    # Create grott.ini file from template
+    if ! sed -e "s|%EXT_URL|$EXT_URL|g" \
```
use only %XXX not $XXX as well and avoid "_" in placeholder names

mstormi commented on functions/grott-install.bash:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    # Create grott.service systemd file from template
+    if ! sed -e "s|%INSTALL_DIR|$INSTALL_DIR|g" \
```
see next comment

andrewfg replied:
see above

andrewfg commented on functions/grott-install.bash:
```diff
@@ -0,0 +1,154 @@
+      echo "Error: Please supply openHAB IP address."
+      echo "Usage: $0 <install|remove> <username> <openhab-ip>"
+      exit 1
+    elif ! [[ "$3" =~ ^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
```
Background is that Grott is a proxy interposed between the Growatt inverter datalogger (on the LAN) and the Growatt cloud server in the cloud. So it passes the data verbatim from the logger to the cloud, and it also forks the same data to the openHAB Growatt binding. So in a generic environment all four functions can be on a different piece of hardware (logger, cloud server, proxy, and openHAB). And this is why in the Growatt readme explains using real IP addresses. However I suppose you are right in that in this case the proxy and openhab will be (must be) on the same machine, so as you say, we could indeed use 127.0.0.1. My only doubt is that I am not 100% sure about the behavior of OH's HTTP server (used in the binding) -- do we know 100% that it always accepts input on 127.0.01? I know it 100% works with a regular dotted ipv4. But maybe there could be use cases where it is blocked for some users?

andrewfg replied:
> isnt it oh talking (opening connections) to grott so rather a question for grott?

The Growatt binding adds a servlet to OH and this receives HTTP POST messages from Grott with the inverter data.

> oh http can be configured to listen on a single ip only or on 0.0.0.0 which is all interfaces.
> In the former case it would not work to use localhost.

Ok. I will revert to us a hard ip address..

andrewfg commented on functions/grott-install.bash:
```diff
@@ -0,0 +1,154 @@
+    echo -n "[openHABian] Installing Grott Proxy with extension URL: $EXT_URL "
+
+    # Update system and install dependencies
+    sudo apt update
```
Yeah. I did wonder about that.

andrewfg commented on functions/grott-install.bash:
```diff
@@ -0,0 +1,152 @@
+
+    # Enable and start service
+    sudo systemctl enable grott --now
+    sudo systemctl daemon-reload
```
Yes I think you are right

andrewfg commented on functions/grott-install.bash:
```diff
@@ -0,0 +1,154 @@
+    # Update system and install dependencies
+    sudo apt update
+    sudo apt install -y python3 python3-pip
+    sudo pip3 install paho-mqtt requests # paho-mqtt is a required dependency (even if disabled)
```
@mstormi I don't understand what you want, so can you please help me? If I understand you correctly you want to call `install_grott "install"` from the command line to see if if works "unattended". This method is one of many helper methods that support the `openhabian-setup.sh` script. When I run `openhabian-setup.sh` without a command line argument, it presents the whiptail dialog 20 | 2F through which it is possible to run `install_grott "install"`. However when I run `openhabian-setup.sh "unattended"` the script also runs, but (obviously) it never calls `install_grott "install"`. So I guess that in the crudest manner I can confirm that it works unattended. By I suppose you want somehow to force `install_grott "install"` to be called from an "unattended" wrapper script. Can you please be specific how you mean that to be done?

andrewfg replied:
A few things..
1. I created the script that you suggested above and I confirm that is runs without error with `install_grott "install"` and `install_grott "remove"`
2. In any case the module is neither a "needed" nor "can't-be-wrong" module so IMHO it should only be called via the UI menu, and never called unattended
3. Anyway the script doesn't use pip3 any more :)
4. And as you say the BATS test also shows it can run unattended

andrewfg replied:
> you might want to auto-install it as part of a fresh install or on deployments with only unattended mode available.

Hmm. The server is only required if the user (will have) installed the Growatt binding. So I think it is probably over-kill to auto-install it on all machines. The service is quite bomb proof and won't cause any errors, or use much resources, if there is not a Growatt logger pushing data to it. However I think we should generally avoid unnecessary wasting of resources.

> Is there any mandatory input? Like the grott server's or oh's ip?

In short -- no:
- The Grott server's ip is `server.growatt.com` which is hardwired into the grott.ini file. I would not suggest making this configurable, since if Growatt would change this, they would anyway be breaking a lot of other things too.
- We dynamically get oh's ip on the fly in the script (as discussed elsewhere in this thread).
- The only other 'variable' is the user name -- which the script inherits from the caller or defaults to `openhabian`.

mstormi replied:
> > you might want to auto-install it as part of a fresh install or on deployments with only unattended mode available.
> 
> Hmm. The server is only required if the user (will have) installed the Growatt binding. So I think it is probably over-kill 
> to auto-install it on all machines. The service is quite bomb proof and won't cause any errors, or use much resources, if 
> there is not a Growatt logger pushing data to it. However I think we should generally avoid unnecessary wasting of
> resources.

I don't mean to install it per default. Only if the users sets that option in openhabian.conf (he will only do so when he runs a Growatt). If there had been some option that needs to be set for install to succeed we could have used that as an indicator that install_grott should run. 
To always also provide the ability for this mode of operation(installation) right with the initial contribution is an openHABian principle we enforce.

andrewfg replied:
> an openHABian principle we enforce

For future reference, are such principles actually written down anywhere?

> Only if the users sets that option in openhabian.conf 

Ok. See my latest commit.

andrewfg commented:
> please check if pip3 also works in unattended

@mstormi apropos this..
1. the module is neither a "needed" nor "can't-be-wrong" module so IMHO it should only be balled via the UI menu, and never called unattended
2. anyway the script doesn't use pip3 any more :)

> don't enlarge the window beyond 25

I reverted the dialog to its prior size (but I added a text "scroll to see more"..)

=> So I think I have now fully addressed all of your issues.

mstormi commented:
> => So I think I have now fully addressed all of your issues. 

Please no checklists type of report. Rather instead go through the review in GitHub and comment per-issue, otherwise it's too hard to keep track of which issue is still current

andrewfg commented:
@mstormi OK I am all finished now (hopefully) .. it passes all CI tests now (except DCO) plus also a live test on an RPi5 64 bit.

andrewfg commented:
@mstormi it is not directly related to this PR but while testing the menu I noticed that there is another whiptail dialog that has a sizing issue..

<img width="1222" height="565" alt="image" src="https://github.com/user-attachments/assets/e99fcb44-6c94-4427-a6ff-7a0ef09f8e5d" />

andrewfg commented:
> fix it while you're on it okay in this pr

Ok. I am happy to fix it here. There are two ways to do it -- 1) widen the dialog, or 2) shorten the texts. => Do you prefer 1 or 2?

andrewfg commented:
> shorten the texts

Ok. Done. It is ready to go again.
